waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -771,6 +771,32 @@
       (catch Throwable t
         (log/error t "Error force-killing pod")))))
 
+(defn signal-service-instance
+  "Sends specified signal to the Kubernetes pod corresponding to the given Waiter Service Instance.
+   Returns nil on success, but throws on failure."
+  [{:keys [api-server-url service-id->service-description-fn] {:keys [force-sigterm-secs sigterm-grace-period-secs]} :pod-bypass :as scheduler}
+   {:keys [service-id] :as instance} service signal-type timeout]
+  (let [pod-url (instance->pod-url api-server-url instance)
+        desc (service-id->service-description-fn service-id)
+        bypass-enabled? (sd/service-description-bypass-enabled? desc)
+        ;; We have to add both of these durations together because K8s gracePeriodSeconds countdown is done
+        ;; in parallel with the preStop hook and issuing the SIGTERM signal.
+        ;; https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination
+        force-sigterm-secs (utils/parse-int (get-in desc ["env" "WAITER_CONFIG_BYPASS_FORCE_SIGTERM_SECS"] (str force-sigterm-secs)))
+        sigterm-grace-period-secs (utils/parse-int (get-in desc ["env" "WAITER_CONFIG_BYPASS_SIGTERM_GRACE_PERIOD_SECS"] (str sigterm-grace-period-secs)))
+        total-bypass-grace-period-secs (+ force-sigterm-secs sigterm-grace-period-secs)
+        grace-period-seconds (if bypass-enabled? total-bypass-grace-period-secs 300)]
+      (log/info "pod-url" pod-url)
+      (case signal-type
+        
+        ; "soft" delete of the pod (i.e., simply transition the pod to "Terminating" state)
+        :sigterm (api-request pod-url scheduler :request-method :delete
+                 :body (utils/clj->json {:kind "DeleteOptions" :apiVersion "v1" :gracePeriodSeconds (or (if (zero? (Integer/parseInt timeout)) nil (Integer/parseInt timeout)) grace-period-seconds)}))
```

Comment:
The function 
```(api-request pod-url scheduler :request-method :delete
                 :body (utils/clj->json {:kind "DeleteOptions" :apiVersion "v1" :gracePeriodSeconds (or (if (zero? (Integer/parseInt timeout)) nil (Integer/parseInt timeout)) grace-period-seconds)}))
```
should be wrapped inside a helper function called `soft-delete-service-instance`